### ISSUE TITLE
Use 100%cpu! work with plotly-extension and 'create a new view of output'  of a chart that output by plotly.

### ISSUE THREAD

**Issue by kondongx** [jupyterlab/jupyterlab#5500]
# Use 100% CPU 

Use 100%cpu! Work with plotly-extension. Right click an output cell -> 'create a new view of output'  of a chart that output by plotly.

## Steps to reproduce the behavior:

you can use this code :
```python
import plotly.offline as py
py.init_notebook_mode(connected=True)
import plotly.graph_objs as go
import plotly.tools as tls

trace = go.Bar(
    x = list(range(1, 5)),
    y = list(range(6, 10))
)
data = [trace]
layout = go.Layout(
    title='My Bar',
)
fig = go.Figure(data=data, layout=layout)
py.iplot(fig)
```

1. copy the above code to a cell
2. ctrl + enter to run this cell
3. right click on the output(a chart) -> click 'create a new view of output'.
4. Then you will find it difficult to scroll the page. And chrome helper process use 100+% cpu(Same in safari).

## Screenshots
<img width="706" alt="2018-10-16 22 10 26" src="https://user-images.githubusercontent.com/28899471/47022631-82843b80-d190-11e8-9362-68a824a27804.png">
<img width="1287" alt="2018-10-16 22 09 44" src="https://user-images.githubusercontent.com/28899471/47022635-857f2c00-d190-11e8-99ce-a8c6096555a4.png">

## Desktop
OS: macOS Mojave 10.14
Browser: Chrome, Safari
JupyterLab: 0.35.1

**Comment by kondongx** [comment]
I had solve this problem by using `plotlywidget` and uninstall `plotly-extension`. 

Here is my solution:

* Read this guide: [Plotly migration guide](https://github.com/plotly/plotly.py/blob/master/migration-guide.md). You will know what is `ploylywidget` doing and what is the different between `plotly offline`.
* Go to [ployly.py-github](https://github.com/plotly/plotly.py/tree/master/plotly), find "JupyterLab Support (Python 3.5+)" section. You can follow this section and install necessary packages **except install plotly-extension**.

My way is a little different from the installation guide. `plotly3.3.0` work fine with jupyterlab0.35.1 on my machine. I did't install jupyterlab0.34 that recommended by the installation guide. And `ipywidgets`, `jupyter-widgets/jupyterlab-manager`, `plotlywidget` are all lasted.

The command that I used as follows:

```bash
jupyter labextension uninstall @jupyterlab/plotly-extension

pip install ipywidgets

# (OS X/Linux :) My machine is mac)
export NODE_OPTIONS=--max-old-space-size=4096
# (Windows)
set NODE_OPTIONS=--max-old-space-size=4096

# Jupyter widgets extension
jupyter labextension install @jupyter-widgets/jupyterlab-manager --no-build

# FigureWidget support
jupyter labextension install plotlywidget --no-build

# Build extensions (must be done to activate extensions since --no-build is used above)
jupyter lab build

# Unset NODE_OPTIONS environment variable
# (OS X/Linux)
unset NODE_OPTIONS
# (Windows)
set NODE_OPTIONS=
```

Done.

**Comment by blink1073** [comment]
Thanks for the recap, glad you got sorted!

**Comment by blink1073** [comment]
cc @gnestor, should we deprecate `plotly-extension` in favor of the upstream soln?

**Comment by dhirschfeld** [comment]
I asked that over at https://github.com/plotly/plotly.py/issues/1054#issuecomment-405064289

> *My current understanding is that `plotly-extension` is not obsolete because it provides the mime-type support needed by `plotly.offline.iplot`, and the support for opening `.plotly.json` files as plots.*
>
> *`plotlywidget` on the other hand provides only the JupyterWidget corresponding to the new `FigureWidget` class.*

**Comment by blink1073** [comment]
Thanks for the background, @dhirschfeld!  @meeseeksdev migrate to @jupyterlab/jupyter-renderers

**Comment by blink1073** [comment]
@meeseeksdev migrate to jupyterlab/jupyter-renderers

**Comment by timkpaine** [comment]
@blink1073 @dhirschfeld no you need both https://github.com/plotly/plotly.py/issues/1071#issuecomment-414851014

**Comment by gnestor** [comment]
I can't reproduce the CPU issue when running this code. 

It's worth mentioning that `py.init_notebook_mode(connected=True)` is unnecessary with plotly-extension installed and I'm 99% certain it doesn't do anything in JupyterLab (because plotly-extension renders the `application/vnd.plotly.v1` output vs. the `text/html` output that inserts plotly.js scripts into the document). See https://github.com/plotly/plotly.py/blob/master/plotly/offline/offline.py#L111-L147 for reference.

**Comment by bollwyvl** [comment]
Haven't looked in a while, but I would wager both of the plotly extensions
could stand to be chunked out at the webpack level with `require.ensure`
(for the time being) or `await import` (In The Future).

This won't help their performance once loaded, but they are both fairly
punishing during the build (2mb monolith):

https://bundlephobia.com/result?p=plotly.js@1.41.3

**Comment by gnestor** [comment]
@bollwyvl Absolutely! It looks like the plotlywidget extension is [6.9MB](https://github.com/plotly/plotly.py/blob/master/plotlywidget/static/index.js)! Could we not take a more general approach and chunk/split out every extension? Also, I assume we're waiting for Firefox and Edge support for dynamic imports before we can start using it?

To those of you who are not familiar with webpack chunking/code splitting, here's an example of it in the vega4-extension: https://github.com/jupyterlab/jupyterlab/blob/8a141cadf904150b01886f189a10a3143676a1eb/packages/vega4-extension/src/index.ts#L186-L210

**Comment by bollwyvl** [comment]
I think the API is basically at the right level now: you want the existence
of your extension known as soon as possible for dependency injection, but
the full weight as late as possible. Since the renderer instance is already
promise-ful, that pretty much fits the bill for most cases.

The worst case I can imagine would be something like a launcher that used
Vega or plotly.

Core is another story. Really only terminal has a unique, big dependency
(xterm). Notebook is just actually a lot of code! Do those need to be
loaded before the animation finishes? Could they even split very well? How
would that jank layout restorer?

We might be able to switch to dynamic import() now that we target ES6.
>

O rly? Will investigate ... Hooray standards from the future past! My first
cut at that for graphviz worked great, but I switched everything to
require.ensure to stay in line with upstream standards.

**Comment by blink1073** [comment]
@bollwyvl, thanks for looking into this!  Do you mean that the feature is in `ts-jest` master but hasn't been released?  Can you point to a PR/issue?

**Comment by blink1073** [comment]
It gets handled at the Webpack level iiuc.

**Comment by vidartf** [comment]
@blink1073 So it should work for projects consuming lab as a library as long as they use webpack as well? Or will additional changes be necessary?

**Comment by blink1073** [comment]
No changes required if using Webpack or an equivalent tool that handles dynamic import (which will eventually include browsers).

**Comment by bollwyvl** [comment]
Right. This isn't landing broadly, but it's in chrome:

https://caniuse.com/#feat=es6-module-dynamic-import

Webpack can consume (and Rollup and parcel require?)  esnext and still
target all the way down to es5. Under the webpack  hood it works the same
as require.ensure, it's just using the standards track syntax, not a
special webpack thing... For now, there are still magic webpack comments
for the chunk name. It's also ten lines shorter, and doesn't require
@types/webpack-env.

Anyhow, esnext-style modules are not suitable for consumption by nodejs
<11 (needs a special extension and a flag...  And I'm staying on 8 until
eol, get off my lawn.), but both Babel and ts can also consume them.

There are also prefetch/preload features, currently supported by webpack,
which would be awesome, but don't currently work with how we patch
publicPath in webpack.config.js with base_url... And might be annoying
anyway, if the moonimation finishes faster, but then the first thing you
want to _do_ is slower because you're loading xxMB of other stuff.

So I guess the right thing to do is start an experiment PR with esnext just
for this renderer, get some tests added to prove it works with jest?

**Comment by blink1073** [comment]
Sounds good!
